lib/internal/modules/esm/module_map.js

```diff
@@ -1,17 +1,85 @@
 'use strict';
 
-const { kImplicitAssertType } = require('internal/modules/esm/assert');
 const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  ArrayPrototypeSort,
+  JSONStringify,
+  ObjectKeys,
   SafeMap,
 } = primordials;
+const { kImplicitAssertType } = require('internal/modules/esm/assert');
 let debug = require('internal/util/debuglog').debuglog('esm', (fn) => {
   debug = fn;
 });
 const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
 const { validateString } = require('internal/validators');
 
-// Tracks the state of the loader-level module cache
-class ModuleMap extends SafeMap {
+/**
+ * Cache the results of the `resolve` step of the module resolution and loading process.
+ * Future resolutions of the same input (specifier, parent URL and import assertions)
+ * must return the same result if the first attempt was successful, per
+ * https://tc39.es/ecma262/#sec-HostLoadImportedModule.
+ */
+class ResolveCache extends SafeMap {
+  constructor(i) { super(i); } // eslint-disable-line no-useless-constructor
+
+  /**
+   * Generates the internal serialized cache key and returns it along the actual cache object.
+   *
+   * It is exposed to allow more efficient read and overwrite a cache entry.
+   * @param {string} specifier
+   * @param {Record<string,string>} importAssertions
+   * @returns {string}
+   */
+  serializeKey(specifier, importAssertions) {
+    // To serialize the ModuleRequest (specifier + list of import attributes),
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
+    // as identical.
+    return specifier + '::' + ArrayPrototypeJoin(
+      ArrayPrototypeMap(
+        ArrayPrototypeSort(ObjectKeys(importAssertions)),
+        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
+      ',');
```

Comment:
Is this faster?

```suggestion
    const keys = ObjectKeys(importAssertions);
    
    if (keys.length === 0) {
      return specifier;
    }

    return specifier + '::' + ArrayPrototypeJoin(
      ArrayPrototypeMap(
        ArrayPrototypeSort(keys),
        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
      ',');
```

Reply:
I'm not sure, it would need to be benchmarked, but more importantly it's making the test fail, so I suggest we leave that to a later PR.

Reply:
You need to update the test so that imports without assertions don’t get a key of `specifier::` but just `specifier`. That was an intentional change in this code, to skip the cost of concatenation and generating a new string.

Reply:
Removing `::` puts us at risk of cache collision, updating the test would not be the solution here, it would hide the fact the implementation is not safe.

Reply:
> Removing `::` puts us at risk of cache collision

How?

Reply:
Well see the test case, I've added it because otherwise it's trivial to create a cache collision.

Reply:
> Well see the test case, I’ve added it because otherwise it’s trivial to create a cache collision.

Because there might be a `::` within the specifier? If that’s a concern, then just add back the concatenation:

```suggestion
    const keys = ObjectKeys(importAssertions);
    
    if (keys.length === 0) {
      return specifier + '::';
    }

    return specifier + '::' + ArrayPrototypeJoin(
      ArrayPrototypeMap(
        ArrayPrototypeSort(keys),
        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
      ',');
```

I would still think that this should be faster than the existing code, as it bypasses the sort and map when there are no assertions.

You might want to consider something other than `::` as the delimiter though, if this is the risk, so that we don’t need to add an unnecessary delimiter for the common case. Perhaps a character that’s disallowed in specifiers, like `\t` or `\n` or something? (Not sure if those are disallowed or what characters are disallowed.)

Reply:
AFAIK any character is allowed in a specifier string. This does feel like premature optimisation, the benchmark didn't show any significant perf regression with the current implementation. Tests are passing, I'd prefer to land that now, and we can do follow-up PRs to optimize further if necessary.

Reply:
I disagree, I feel like there's no way this isn't faster and I've said for a while now that we should optimize for the common case of no assertions. I'm happy to rerun CI for you to get it green again.

Reply:
If we land as is, and open a follow up PR, it will be trivial to compute if your proposal is indeed better for perf and how much.

Reply:
Even if it doesn't improve the perf noticeably, I like @GeoffreyBooth's suggestion because it makes the code easier to understand. It becomes obvious that the complex part is here to deal with import assertions.